### ISSUE TITLE
[HOLD for payment 2023-09-06] [$1000] mweb - Inconsistency: `Copy Link` context menu item does not show on the attachment message when it is opened over the attachment in mweb|web

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#21346]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Open a DM with any user in NewDot.
2. Send a attachment message.
3. Long press over the attachment to open the context menu.

## Expected Result:
On mweb, it should show `copy link` item as well when opened over the attachment same like native application. 
We should either shown for both mWeb and native or hide on both. I think better would be to show it as there is less space for the user to open the full context menu for attachment(which is opened when user long presses the empty space side to the attachment).

## Actual Result:
On mweb, `copy link` menu item is not shown when menu is opened over the attachment. This is inconsistent with the native iOS and Android.

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [x] Android / Chrome
- [ ] iOS / native
- [x] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.28-3

**Reproducible in staging?:** y

**Reproducible in production?:** y

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/93399543/90ef3010-6f86-4fea-bfe0-6e7f174b56ed

![Screenshot_20230622_154047_Chrome](https://github.com/Expensify/App/assets/93399543/a5736b79-3075-4ce2-9a70-1ab2c0dee777)

**Expensify/Expensify Issue URL:**

**Issue reported by:** @parasharrajat 

**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1686904574870779

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0152b793902d18fd0b</li>
        <li>Upwork Job ID: 1673348174719672320</li>
        <li>Last Price Increase: 2023-07-17</li>
    </ul>
</details>

**Comment by jeet-dhandha** [comment]
@kbecciv @flaviadefaria - This was disabled around this,
PR: https://github.com/Expensify/App/pull/9955
Issue: https://github.com/Expensify/App/issues/9660.
We were waiting to add a copy image to clipboard functionality but we aren't there yet probably.

If we want to be at a situation where user can identify between copy link of message and not mistake it for copy link of image. Then we can change `Copy Link` to `Copy Message Link` and remove the `isAttachment` check to show the `Copy Message Link` again.

**Comment by s-alves10** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Copy link menu item is not shown when menu is opened over the attachment on mobile web unlike mobile native application

### What is the root cause of that problem?

We have image renderer showing context menu here
https://github.com/Expensify/App/blob/6183609c577f2310928bb4e8ee23b52f6dd28e00/src/components/HTMLEngineProvider/HTMLRenderers/ImageRenderer.js#L68
Anchor element here is `img` element and we intentionally prevent `Copy Link` menu here
https://github.com/Expensify/App/blob/6183609c577f2310928bb4e8ee23b52f6dd28e00/src/pages/home/report/ContextMenu/ContextMenuActions.js#L202-L206

As you can see, when the message is attachment and anchor is `img`, we don't show `Copy Link`. For native applications, `img` element is translated into React Native View by `ImageRenderer` and doesn't meet this condition. Because of this, Copy Link doesn't appear in context menu on mobile web unlike mobile native applications
I'm not sure this is a bug. Anyway this is the root cause

### What changes do you think we should make in order to solve the problem?

We should remove checking `img` attchement checking. `shouldShow` would be like

```
        shouldShow: (type, reportAction, isArchivedRoom, betas, menuTarget) => {
            return Permissions.canUseCommentLinking(betas) && type === CONTEXT_MENU_TYPES.REPORT_ACTION && !ReportActionUtils.isMessageDeleted(reportAction);
        },
```

This works as expected.

<details>
    <summary>Result</summary>


</details>

### What alternative solutions did you explore? (Optional)

**Comment by jeet-dhandha** [comment]
@parasharrajat Are we expecting `copy link` over here or are we still waiting to get a `copy image` and then allow it on image press ?

**Comment by flaviadefaria** [comment]
Added the external label. I agree with making this consistent for mWeb.

**Comment by jeet-dhandha** [comment]
@flaviadefaria - Any thoughts on this comment https://github.com/Expensify/App/issues/21346#issuecomment-1603318986 ?

**Comment by flaviadefaria** [comment]
@jeet-dhandha I agree with you that adding `copy link` doesn't make sense. I actually posted [here](https://expensify.slack.com/archives/C049HHMV9SM/p1686904574870779) to understand a bit better the inconsistency reported as I couldn't find it.

**Comment by s-alves10** [comment]
@0xmiroslav 

Can you explain again what is the expected behavior of this issue?

**Comment by s-alves10** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/21346#issuecomment-1604028149)

**Comment by jeet-dhandha** [comment]
@0xmiroslav @parasharrajat  - Should i create a PR for this ?

**Comment by jeet-dhandha** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
- Copy Link options is inconsistent with the rest of platforms. Either it should be removed or it should be added to the rest of platforms.

### What is the root cause of that problem?
- The root cause is that condition isAttachmentTarget is not true for all platforms. As IMG tag is not supported in all platforms.

### What changes do you think we should make in order to solve the problem?
- We should remove the Copy Link option from the menu for all platforms, if its an attachment.
- Code wise we just need to remove `isAttachmentTarget` condition from the code and instead use `isAttachment`.

### What alternative solutions did you explore? (Optional)
- N/A

PS: I didn't post any proposal as I was not sure of the requirement due to previous open issues.

**Comment by flaviadefaria** [comment]
@0xmiroslav and thoughts on @jeet-dhandha's proposal above?

**Comment by s-alves10** [comment]
@0xmiroslav 

Please check the updated proposal
https://github.com/Expensify/App/issues/21346#issuecomment-1604028149
https://github.com/Expensify/App/pull/22315

**Comment by jeet-dhandha** [comment]
@0xmiroslav May I kindly ask what your expectations are for the solution?

**Comment by s-alves10** [comment]
@0xmiroslav 

I thoroughly checked the project related to eh context menu and verified that manual adding doesn't have any side effect.
This also sticks to the minimum changes rule of the proposal, I think. Anyway, let me try to find the better solution

**Comment by jeet-dhandha** [comment]
https://github.com/Expensify/App/issues/21346#issuecomment-1624215407 @0xmiroslav my proposal haven’t introduced any tagName param or prop. Will you give it a look again ?

**Comment by jeet-dhandha** [comment]
Subscribed. Lets wait on what thoughts Rajat has on this.

**Comment by flaviadefaria** [comment]
Do we have an update here?

**Comment by flaviadefaria** [comment]
@0xmiroslav I'm struggling to recreate this. What does clicking `copy link` actually copy? Unless it copies the link to the message then I think it's fine to remove it.

**Comment by flaviadefaria** [comment]
Are you still able to recreate this in iOS native? I can no longer see the `copy link` option there.

**Comment by flaviadefaria** [comment]
@0xmiroslav I enabled the beta on my account and I still can't reproduce this, but if you can then that's fine. Can you reply to this question:

> What does clicking copy link actually copy? Unless it copies the link to the message then I think it's fine to remove it.

**Comment by flaviadefaria** [comment]
Confirmed in Slack. @0xmiroslav let's go with what you proposed above and only have the clickable area display the "Copy link" option linking to the message, not the image.

**Comment by s-alves10** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/21346#issuecomment-1604028149)

**Comment by flaviadefaria** [comment]
@0xmiroslav can you please give an update?

**Comment by situchan** [comment]
I'd like to take this as C+ based on [discussion](https://expensify.slack.com/archives/C02NK2DQWUX/p1691641008846119?thread_ts=1691488353.531539&cid=C02NK2DQWUX)

**Comment by flaviadefaria** [comment]
Thanks @situchan I just reassigned it to you.

**Comment by situchan** [comment]
@jeet-dhandha is the first to provide solution which meets expected result (_hide Copy link on attachment for all platforms_) we finally landed on.
Proposal: https://github.com/Expensify/App/issues/21346#issuecomment-1624215407

🎀 👀 🎀 C+ reviewed

**Comment by situchan** [comment]
> The menu item should be shown on attachment messages but not over the attachment view on mobile.

Exactly. I accepted that solution. Still show option on message itself but not on attachment view, no matter mobile or web.
https://github.com/Expensify/App/issues/21346#issuecomment-1653591076

**Comment by jeet-dhandha** [comment]
I will update my proposal with above requirements. @situchan @parasharrajat

**Comment by situchan** [comment]
I do apologize for everyone. I got confused about the expected result, so updated previous approval [comment](https://github.com/Expensify/App/issues/21346#issuecomment-1673067654).

Now re-evaluating proposals

**Comment by situchan** [comment]
@s-alves10 thanks for updated proposal. The root cause is correct. Your solution will override mWeb's nativeEvent as well, which is redundant.

**Comment by s-alves10** [comment]
Yeah, we can avoid that

**Comment by s-alves10** [comment]
We do tagName check only in `Copy Link` menu and only check the `target.tagName`
So we can simply use `{nativeEvent: {target: {tagName: 'IMG'}}}`

**Comment by situchan** [comment]
@s-alves10 why can't we fully customize it like `{target: 'image'}`? And no need to pass full `nativeEvent` as it's not used anywhere.

**Comment by jeet-dhandha** [comment]
@situchan We are using it here so we need to pass the nativeEvent but we can override the `target` with `{target: 'IMG'}` and change isAttachmentTarget condition to below.
- `const isAttachmentTarget = menuTarget === 'IMG' && isAttachment;`
- https://github.com/Expensify/App/blob/7d59bbb4a740a267c0a17a0ab7f369c0bd269dec/src/pages/home/report/ContextMenu/PopoverReportActionContextMenu.js#L163-L170

**Comment by situchan** [comment]
ok so 👇 is wrong

> We do tagName check only in `Copy Link` menu and only check the `target.tagName` So we can simply use `{nativeEvent: {target: {tagName: 'IMG'}}}`

**Comment by jeet-dhandha** [comment]
@situchan Are you asking that the comment is wrong or telling me ?

**Comment by situchan** [comment]
> Please do update the C+ review whenever possible.

are you expecting that your proposal will be accepted?

@s-alves10 is the first to propose working solution which fixes the root cause.

**Comment by jeet-dhandha** [comment]
@situchan No. This wasn't in my luck 🍀 to get accepted, so what i can do is help others achieve better results and make them get better payout by minimising the chance of regression by their PR.

**Comment by jeet-dhandha** [comment]
I am but a humble soul helping Expensify Bloom 🌷

**Comment by s-alves10** [comment]
> @s-alves10 why can't we fully customize it like `{target: 'image'}`? And no need to pass full `nativeEvent` as it's not used anywhere.

I wanted to fix the issue with minimal changes. We're using `nativeEvent` [here](https://github.com/Expensify/App/blob/7d59bbb4a740a267c0a17a0ab7f369c0bd269dec/src/pages/home/report/ContextMenu/PopoverReportActionContextMenu.js#L163-L170) for position calculations

**Comment by situchan** [comment]
> @situchan No. This wasn't in my luck 🍀 to get accepted, so what i can do is help others achieve better results and make them get better payout by minimising the chance of regression by their PR.

@jeet-dhandha thanks for your help. Of course such things will be handled in PR.

> I think the current solution is fine because mweb has this structure of data already. What do you think?

yes, fine to me too

**Comment by situchan** [comment]
@s-alves10's [proposal](https://github.com/Expensify/App/issues/21346#issuecomment-1604028149) LGTM!
cc: @youssef-lr 
🎀 👀 🎀 C+ reviewed

**Comment by flaviadefaria** [comment]
Looks like we're just waiting on @youssef-lr for the review.

**Comment by flaviadefaria** [comment]
@laurenreidexpensify I'm heading OoO for 1 week so assigning this to you while I'm away.

**Comment by laurenreidexpensify** [comment]
@youssef-lr bump for update thanks

**Comment by youssef-lr** [comment]
LGTM as well! Assigned.

**Comment by s-alves10** [comment]
@situchan 

[PR](https://github.com/Expensify/App/pull/25700) is ready for review

**Comment by s-alves10** [comment]
@youssef-lr 

Please take a loot at the [PR](https://github.com/Expensify/App/pull/25700)

**Comment by situchan** [comment]
BZ Checklist:
No regression. This is inconsistency, not bug. As this is still beta feature, regression test is not needed for now.

**Comment by situchan** [comment]
> > As this is still a beta feature, a regression test is not needed for now.
> 
> I don't think we should ignore the steps due to this reason. Or do you have any reference to some discussion where you confirmed this?

I don't have any reference. But this doesn't affect production users yet.
I don't think it's late to add regression test step after message deep link is implemented and beta is removed.

**Comment by flaviadefaria** [comment]
>For reference, here are some details about the assignees on this issue:
@situchan requires payment
@s-alves10 requires payment 
@parasharrajat does not require payment (Eligible for Manual Requests)

Payment Summary:

@situchan requires payment  = $1000 (no bonus)
@s-alves10 requires payment = $1000 (no bonus)
@parasharrajat does not require payment (Eligible for Manual Requests) = $250 (reporting bonus)

**Comment by situchan** [comment]
@flaviadefaria I think this is eligible for #urgency bonus.
- [Assign date](https://github.com/Expensify/App/issues/21346#event-10158594734): Aug 22
- [C+ approved](https://github.com/Expensify/App/pull/25700#pullrequestreview-1593996583) and requested engineer review: Aug 24
- [Merged without change requests](https://github.com/Expensify/App/pull/25700#event-10198244130): Aug 26

**Comment by flaviadefaria** [comment]
@situchan @s-alves10 offers sent to both of you.

**Comment by flaviadefaria** [comment]
That's fair @situchan I'll add the 50% bonus once you've both accepted the contract.

New Payment Summary:

@situchan requires payment = $1000 + $500 (bonus) = $1500
@s-alves10 requires payment  = $1000 + $500 (bonus) = $1500
@parasharrajat does not require payment (Eligible for Manual Requests) = $250 (reporting bonus)

**Comment by flaviadefaria** [comment]
@situchan I'm waiting for you to accept the offer
@s-alves10 has been paid

**Comment by flaviadefaria** [comment]
Everyone has been paid.

**Comment by flaviadefaria** [comment]
Oh sorry that I closed this before you were paid @parasharrajat! I'll reopen and switch this to weekly.

**Comment by JmillsExpensify** [comment]
$250 payment for @parasharrajat approved based on BZ summary.

**Comment by youssef-lr** [comment]
Are we good to close this out?

**Comment by flaviadefaria** [comment]
Alright seems like we're all done here - closing it!
